File: src/lib/password-rules.ts
export interface PasswordRuleContext {
  // Contraseña temporal o actual del usuario, cuando corresponde
  // (crear usuario, resetear contraseña) — la nueva no puede ser igual.
  passwordActual?: string;
}

export interface PasswordRule {
  id: string;
  label: string;
  test: (password: string, context?: PasswordRuleContext) => boolean;
}

// Lista mínima de contraseñas genéricas/comunes bloqueadas, además de la
// comparación contra la contraseña temporal/actual (vía el contexto).
const CONTRASENAS_GENERICAS = [
  'password123!',
  'password1234',
  '123456789!',
  'qwerty123!',
  'admin1234!',
  'devtech123!',
  'bienvenido1!',
];

export const passwordRules: PasswordRule[] = [
  {
    id: 'length',
    label: 'Mínimo 10 caracteres',
    test: (password) => password.length >= 10,
  },
  {
    id: 'uppercase',
    label: 'Al menos una mayúscula (A-Z)',
    test: (password) => /[A-ZÁÉÍÓÚÑ]/.test(password),
  },
  {
    id: 'lowercase',
    label: 'Al menos una minúscula (a-z)',
    test: (password) => /[a-záéíóúñ]/.test(password),
  },
  {
    id: 'number',
    label: 'Al menos un número (0-9)',
    test: (password) => /[0-9]/.test(password),
  },
  {
    id: 'special',
    label: 'Al menos un carácter especial (!@#$...)',
    test: (password) => /[^A-Za-z0-9]/.test(password),
  },
  {
    id: 'not-generic',
    label: 'No ser una contraseña genérica o la temporal recibida',
    test: (password, context) => {
      if (CONTRASENAS_GENERICAS.includes(password.toLowerCase())) {
        return false;
      }
      if (context?.passwordActual && password === context.passwordActual) {
        return false;
      }
      return true;
    },
  },
];

export interface PasswordRuleResult {
  id: string;
  label: string;
  met: boolean;
}

export function evaluatePasswordRules(
  password: string,
  context?: PasswordRuleContext,
): PasswordRuleResult[] {
  return passwordRules.map((rule) => ({
    id: rule.id,
    label: rule.label,
    met: rule.test(password, context),
  }));
}

export function isPasswordValid(
  password: string,
  context?: PasswordRuleContext,
): boolean {
  return passwordRules.every((rule) => rule.test(password, context));
}

File: OwnSpaceAPI/src/OwnSpaceAPI.Api/Services/PasswordRules.cs
using System.Text.RegularExpressions;

namespace OwnSpaceAPI.Api.Services;

public sealed class PasswordRuleContext
{
  //Contraseña temporal del usuario.
  // La nueva no puede ser igual luego de haber realizado el cambio
  public string? PasswordActual {get; init; }
}
public sealed record PasswordRule(string Id, string Label, Func<string, PasswordRuleContext?, bool> Test);
public sealed record PasswordRuleResult(string Id, string Label, bool Met);

public static class PasswordRules
{
  // Duplicada a mano en src/lib/password-rules.ts del frontend (esta
  // lista de acá es la autoridad real; la del frontend solo da feedback
  // en vivo en el formulario) — mantener las dos listas iguales si se
  // edita una. "12345" se sacó: nunca hizo nada, 5 caracteres ya
  // rechaza por la regla de longitud mínima antes de llegar acá.
  private static readonly HashSet<string> ContrasenasGenericas = new(StringComparer.OrdinalIgnoreCase)
    {
        "password123!",
        "password1234",
        "123456789!",
        "qwerty123!",
        "admin1234!",
        "devtech123!",
        "bienvenido1!",
        "contrasena",
        "contraseña",
        "incorrecta",
        "incorrecto",
    };

    public static readonly IReadOnlyList<PasswordRule> Rules = new List<PasswordRule>
    {
        new("length", "Mínimo 10 caracteres",
            (p, _) => p.Length >= 10),

        new("uppercase", "Al menos una mayúscula (A-Z)",
            (p, _) => Regex.IsMatch(p, "[A-ZÁÉÍÓÚÑ]")),

        new("lowercase", "Al menos una minúscula (a-z)",
            (p, _) => Regex.IsMatch(p, "[a-záéíóúñ]")),

        new("number", "Al menos un número (0-9)",
            (p, _) => Regex.IsMatch(p, "[0-9]")),

        new("special", "Al menos un carácter especial (!@#$...)",
            (p, _) => Regex.IsMatch(p, "[^A-Za-z0-9]")),

        new("not-generic", "No ser una contraseña genérica o la temporal recibida",
            (p, ctx) =>
            {
                if (ContrasenasGenericas.Contains(p))
                {
                    return false;
                }
                if (!string.IsNullOrEmpty(ctx?.PasswordActual) && p == ctx.PasswordActual)
                {
                    return false;
                }
                return true;
            }),
    };
    public static IReadOnlyList<PasswordRuleResult> Evaluate(string password, PasswordRuleContext? context = null) =>
        Rules.Select(r => new PasswordRuleResult(r.Id, r.Label, r.Test(password, context))).ToList();

    public static bool IsValid(string password, PasswordRuleContext? context = null) =>
        Rules.All(r => r.Test(password, context));



}